consensus/src/consensus/remote_data_store.rs

```diff
@@ -48,6 +51,62 @@ enum RemoteDataStoreOps {
 }
 
 impl<N: Network> RemoteDataStore<N> {
+    const MAX_BLOCK_DRIFT: usize = 10;
+
+    /// Waits for the block with given `block_hash` to be encountered and returns it. Will only wait for
+    /// [RemoteDataStore::MAX_BLOCK_DRIFT] events that add blocks.
+    ///
+    /// Returns None if the block cannot be retrieved even after waiting
+    async fn get_or_await_block(
+        blockchain: &BlockchainProxy,
+        block_hash: &Blake2bHash,
+    ) -> Option<Block> {
+        // Either get the block if it exists yet, or get the notifier stream.
+        // Blockchain must be dropped here, as it will be held across an await otherwise.
+        let blockchain_event_stream = {
+            let blockchain = blockchain.read();
+            // First try to obtain, from our chain store, the block that was used to generate the proof
+            if let Ok(block) = blockchain.get_block(block_hash, false) {
+                return Some(block);
+            }
+            blockchain.notifier_as_stream()
+        };
+
+        // Wait for MAX_BLOCK_DRIFT events that add blocks, while checking if the added blocks contain the
+        // block hash in question
+        let item = blockchain_event_stream
+            .filter(|event| std::future::ready(!event.added_hashes().is_empty()))
+            .take(Self::MAX_BLOCK_DRIFT)
+            .filter(|event| std::future::ready(event.added_hashes().contains(&block_hash)))
```

Comment:
Avoid calling `added_hashes` twice:
```suggestion
            .filter_map(async |event|  { let hashes = event.added_hashes(); if hashes.is_empty() { None } else { Some(hashes) }))
            .take(Self::MAX_BLOCK_DRIFT)
            .filter(|hashes| hashes.contains(block_hash)))
```